src/Language/PureScript/Sugar/LetPattern.hs

```diff
@@ -0,0 +1,48 @@
+-- |
+-- This module implements the desugaring pass which replaces patterns in let-in
+-- expressions with appropriate case expressions.
+--
+module Language.PureScript.Sugar.LetPattern (desugarLetPatternModule) where
+
+import Prelude.Compat
+
+import Control.Monad.Error.Class (MonadError(..))
+import Control.Monad.Supply.Class
+
+import Language.PureScript.AST
+import Language.PureScript.Errors
+
+-- |
+-- Replace every @BoundValueDeclaration@ in @Let@ expressions with @Case@
+-- expressions.
+--
+desugarLetPatternModule :: forall m. (MonadSupply m, MonadError MultipleErrors m) => Module -> m Module
+desugarLetPatternModule (Module ss coms mn ds exts) = Module ss coms mn <$> parU ds desugarLetPattern <*> pure exts
+
+-- |
+-- Desugar a single let expression
+--
+desugarLetPattern :: forall m. (MonadSupply m, MonadError MultipleErrors m) => Declaration -> m Declaration
+desugarLetPattern (PositionedDeclaration pos com d) = PositionedDeclaration pos com <$> rethrowWithPosition pos (desugarLetPattern d)
+desugarLetPattern decl =
+  let (f, _, _) = everywhereOnValuesM return replace return
+  in f decl
+  where
+  replace :: Expr -> m Expr
+  replace (Let ds e) = go ds [] e
+  replace other = return other
+
+  go :: [Declaration] -- ^ Declarations to desugar
+        -> [Declaration] -- ^ Declarations already desugared, in reverse order
```

Comment:
Nitpick: can you please line up the `->` and `::`s and move the comments underneath?